flink-connector-opensearch/src/test/java/org/apache/flink/streaming/connectors/opensearch/OpensearchSinkTest.java

```diff
@@ -0,0 +1,562 @@
+        builder.setFailureHandler(new NoOpFailureHandler());
+
+        final OpensearchSink<String> sink = builder.build();
+        final OneInputStreamOperatorTestHarness<String, Object> testHarness =
+                new OneInputStreamOperatorTestHarness<>(new StreamSink<>(sink));
+
+        testHarness.open();
+
+        responses.add(
+                createResponse(
+                        new BulkItemResponse(
+                                1,
+                                OpType.INDEX,
+                                new Failure(
+                                        "test",
+                                        "_doc",
+                                        "1",
+                                        new Exception("artificial failure for record")))));
+        testHarness.processElement(new StreamRecord<>("msg"));
+
+        assertThatThrownBy(() -> testHarness.snapshot(1L, 1000L))
+                .getCause()
+                .getCause()
+                .hasMessageContaining("artificial failure for record");
+    }
+
+    /**
+     * Tests that any item failure in the listener callbacks due to flushing on an immediately
+     * following checkpoint is rethrown; we set a timeout because the test will not finish if the
+     * logic is broken.
+     */
+    @Test
+    @Timeout(5)
+    public void testItemFailureRethrownOnCheckpointAfterFlush() throws Throwable {
+        final OpensearchSink.Builder<String> builder =
+                new OpensearchSink.Builder<>(
+                        Arrays.asList(new HttpHost("localhost", server.getLocalPort())),
+                        new SimpleSinkFunction<String>());
+        builder.setBulkFlushInterval(1000);
+        builder.setFailureHandler(new NoOpFailureHandler());
+
+        final OpensearchSink<String> sink = builder.build();
+        final OneInputStreamOperatorTestHarness<String, Object> testHarness =
+                new OneInputStreamOperatorTestHarness<>(new StreamSink<>(sink));
+
+        testHarness.open();
+
+        responses.add(
+                createResponse(
+                        new BulkItemResponse(
+                                1,
+                                OpType.INDEX,
+                                new IndexResponse(
+                                        new ShardId("test", "-", 0), "_doc", "1", 0, 0, 1, true))));
+
+        responses.add(
+                createResponse(
+                        new BulkItemResponse(
+                                2,
+                                OpType.INDEX,
+                                new Failure(
+                                        "test",
+                                        "_doc",
+                                        "2",
+                                        new Exception("artificial failure for record")))));
+
+        testHarness.processElement(new StreamRecord<>("msg-1"));
+
+        // Await for flush to be complete
+        awaitForFlushToFinish(1);
```

Comment:
You can avoid this kind of busy waiting quite easily by using `OneShotLatches`.

```
    private OneShotLatch addResponse(Consumer<HttpResponse> consumer) {
        OneShotLatch oneShotLatch = new OneShotLatch();
        responses.add(
                response -> {
                    consumer.accept(response);
                    oneShotLatch.trigger();
                });
        return oneShotLatch;
    }
....
      OneShotLatch firstResponse = responses.add(
                createResponse(
                        new BulkItemResponse(
                                1,
                                OpType.INDEX,
                                new IndexResponse(
                                        new ShardId("test", "-", 0), "_doc", "1", 0, 0, 1, true))));
...
    firstResponse.await();
```

That being said, this whole "wait for server response to be consumed" approach is a bit sketchy?
We're only waiting for the server to send the response, _not_ for the connector to process it.

Reply:
I would argue there is definitely no need for additional latch: the queue is the "latch" by itself - we know when it is drained. We could also use https://github.com/awaitility/awaitility to make the waiting logic clean.

> That being said, this whole "wait for server response to be consumed" approach is a bit sketchy?

We are dealing with async system so I think this is kind of expectable: we are waiting for some events to happen at some point (be it latch, queue or other means). The sink does not provide any means to hook inside and, for example, we could observe the "process it" part using `getNumPendingRequests` only.

Reply:
> The point was to not wait by looping over a condition to save CI resources. 

Sure, replaced with `Condition`, should be simpler than per-response monitors.

> This is especially a problem when the condition is never fulfilled and we just churn through CPU cycles for no good reason.

All tests are timeboxed to 5 seconds

Reply:
> All tests are timeboxed to 5 seconds

a) In our experience such small timeouts actually cause tests to be unstable.
b) code has a tendency to travel. Arguing that "this code is fine because of that code _over there_" basically means "this code is unsafe to copy", which isn't a good state to be in.

Reply:
The tests seem to be very stable, I was running them continuously for 2h and have not seen any flakyness.

Reply:
It's more of a CI problem. Time jumping ahead and things like that. :weary:

Reply:
I think we could always update the test cases, if the instability comes in, the 5 seconds buffer is 3x buffer for what test case should take.